README.md
# Houston :: A static files server based on node.js
 
Yet another static file server, simple as you hear.

The main purpose of Houston is to provide a static, 0 dependencies (just mime), server. 

**WARNING** Houston 1.0.x only runs on node-v0.8.x. The 0.6.x compatible version is 0.2.1 (`npm install -g nhouston@0.2.1`)

## Features

Houston initially was made to provide a cli-server. But the ability to create cli apps with custom servers make me re-think about this. Still Houston is a cli-server but with the followed features in order to you:

-  static file server
-  spdy, https support
-  houston is an instance of a server, so you can tweak an put socket.io in from of it
-  `cli: true` enable cli features like process.argv parser and more -> (houston#config)
-  websockets support (see socket.io example)
- If Houston can't handle the event it'll let you the task. (e.g. `houston.on('POST', function(req, res){/* the response code */}))`


## Installation

Asuming that you have installed [node.js](http://nodejs.org) and [npm](http://npmjs.org)
    
    > npm install -g nhouston

Or:

    > git clone https://github.com/alejandromg/Houston.git
    > cd Houston
    > ln -s bin/houston ~/bin
    

Then:
   
    # To start the server in the current dir 
    > houston 

This will open a new window in your browser, or go to [http://localhost:8000/](http://localhost:8000/)

To start the server in a different port:

    > sudo houston --port 80
    > houston -p 8010
    > houston -p=8010

All of them are valid, also you can define a different path:

    > houston --path=/home/ --port=8010

Also you can avoid the new browser window with:

    > houston -b false

## Static server

Houston, as well, has a static file server built-in. So if you want to use Houston as your static files provider use it as follows:

    var houston = require('nhouston');
    var server = houston(options);

Where options can be:

    { 
      path : __dirname + '/public', // to serve files from public folder
      port : 3001, // the port that houston will bind to listen
      spdy : { // To create an spdy server (npm install spdy), this can be https too
        key: fs.readFileSync(__dirname + '/keys/spdy-key.pem'),
        cert: fs.readFileSync(__dirname + '/keys/spdy-cert.pem'),
        ca: fs.readFileSync(__dirname + '/keys/spdy-csr.pem')
      },
      silent : true, // Log options: false as default
      cli    : true, // to emule cli behavior (list files on dirs, open browser and more)
    }

See more in the examples dir.

## Screenshot

<img src="http://dl.dropbox.com/u/29662133/Pantallazo-Houston%20%3A%3A%20-home-alejandromg-dev-blackbox%20-%20Google%20Chrome.png" />

## Help

`houston -h`
`houston --help`

    Houston :: A cool static files server

    Available options:
       --port, -p    Listening port to Houston, default to 8000 
       --path, -d    Dir of starting point to Houston, default to actual dir
       --browser,-b  open browser window, (true,false) default to true
       --help, -h    show this info
       --version,-v  Show the current version of Houston

    :: end of help ::

## Alternatives

For sure that Houston is not the first option to static file servers, there are a bunch of other static file providers, with better support:

-  [ecstatic](https://github.com/jesusabdullah/node-ecstatic)
-  connect static middleware

And many others.

## Contributors

- [Alejadro Morales](http://github.com/alejandromg)
- [Kronos](http://github.com/alejandromg/kronos 'npm install kronos') 

## License

MIT 2012

History.md

1.0.1 / 2012-07-07 
==================

  * A better font-size on header 
  * No more 404 on URL with spaces 
  * Fix Content-type on 404 response 
  * More examples 
  * New Font "Pontano Sans" 
  * New Option Key: 'ignorePort' (View example `simple')

1.0.0 / 2012-07-07 
==================

  *  Static Server 
  *  SPDY, HTTPS
  *  websockets support 
  *  More customizable 
  *  Faster (10x)
  *  Examples 
  *  Propertly handles 304 codes (ETag) 
   
n.n.n / 2012-04-05 
==================

  * [fix] support for relative paths like: ~/dev
  * [dist] Fixed screenshot
  * [dist] Added screenshot and houston version visible
  * [dist] Added houston version in footer
  * [dist] bump to 0.2.0
  * [dist] Added streams better
  * [dist] Changelog for 0.1.3
  * [dist] Bump to 0.1.3 and browser fix
  * [dist] Fixed chmod and bump to 0.1.2
  * [dist] Code clean; bump to 0.1.1
  * [dist] First prototype ready
  * [kronos] Initial Commit

n.n.n / 2012-04-05 
==================

  * [dist] Added houston version in footer
  * [dist] bump to 0.2.0
  * [dist] Added streams better
  * [dist] Changelog for 0.1.3
  * [dist] Bump to 0.1.3 and browser fix
  * [dist] Fixed chmod and bump to 0.1.2
  * [dist] Code clean; bump to 0.1.1
  * [dist] First prototype ready
  * [kronos] Initial Commit

n.n.n / 2012-04-04 
==================

  * [dist] Changelog for 0.1.3
  * [dist] Bump to 0.1.3 and browser fix
  * [dist] Fixed chmod and bump to 0.1.2
  * [dist] Code clean; bump to 0.1.1
  * [dist] First prototype ready
  * [kronos] Initial Commit

n.n.n / 2012-03-29 
==================

  * [dist] Bump to 0.1.3 and browser fix
  * [dist] Fixed chmod and bump to 0.1.2
  * [dist] Code clean; bump to 0.1.1
  * [dist] First prototype ready
  * [kronos] Initial Commit


lib/server.js
#!/usr/bin/env node

/*
 * Houston
 * @date Sun Mar 25 2012 16:06:16 GMT-0600 (CST)
 * @author Alejandro Morales <vamg008@gmail.com>
 * @name server.js
 * @licence MIT
 * 
 */

/*jshint node:true, noempty:true, laxcomma:true, nonstandard:true, laxbreak:true */

"use strict";

var domain  = require('domain')
  , http    = require('http')
  , https   = require('https')
  , fs      = require('fs')
  , path    = require('path')
  , url     = require('url')
  , qs      = require('querystring')
  , mime    = require('mime')
  , colors  = require('colors')
  , version = require(__dirname + '/../package.json').version
  , runningTime = Date.now()
  , existsSync  = fs.existsSync || path.existsSync
  , params  = {}
  , template, csl
  ;

// Default configuration

var cfg = {
  port : 8000,
  path : process.env.PWD,
  browser : false
};

function load () {
  // 
  // Only launched on cli mode
  // Get the params
  // Do it right (?) so JSHint  can't complain
  //
  var reg = new RegExp('--')
    , sir = new RegExp('-')
    , par = new RegExp('=.*','gi');

  // Cli config
  cfg.path = cfg.path || process.env.PWD;
  cfg.safe = cfg.safe || 'false';
  cfg.verbose = cfg.verbose || true;
  cfg.browser = cfg.browser || true;

  for (var key in process.argv){
    var attr = process.argv[key];

    if (/--/gi.test(attr) || /-/gi.test(attr)){
      var attri = attr.replace(reg,'')
                      .replace(sir,'')
                      .replace(par,'');
      if (par.test(attr)){
        params[attri] = attr.split('=')[1];
      } else {
        params[attri] = process.argv[++key];
      }
    }
  }
  // Default configuration

  // Help function
  function logHelp () {
    console.log(' ');
    console.log(' Houston :: A cool static files server');
    console.log(' ');
    console.log(' Available options:');
    console.log('\t --port, -p \t Listening port to Houston, default to 8000 ');
    console.log('\t --path, -d \t Dir of starting point to Houston, default to actual dir');
    console.log('\t --browser,-b \t open browser window, (true,false) default to true');
    console.log('\t --help, -h \t show this info');
    console.log('\t --version,-v \t Show the current version of Houston');
    console.log(' ');
    console.log(' :: end of help ::');
    process.kill(0);
  }

  Object.keys(params).map(function(key){
    switch (key){
      case 'port':
      case 'p':
        cfg.port = params[key];
        break;
      case 'path':
      case 'd':
        if (params[key] == '.' || !params[key])
          cfg.path = process.env.PWD;
        else 
          cfg.path = params[key];
        if (cfg.path.charAt(0) == '~')
          cfg.path = path.join(process.env.HOME, cfg.path.replace(/\~/,''));
        break;
      case 'browser':
      case 'b':
        cfg.browser = params[key];
        break;
      case 'v':
      case 'version':
        console.log('Houston '+ version);
        process.kill(0);
        break;
      case 'help':
      case 'h':
        logHelp();
        break;
      case 'silent':
      case 's':
        cfg.verbose = true;
        break;
      default:
       break;
    }
  });

  template = fs.readFileSync(path.join(__dirname, '..', 'public/index.html'), 'utf8');
}

// dummy (awful) logger 
function Logger (){
  var logo = '⢷⡾ Houston ➔';

  // Logos
  var s = logo.bold.green
    , s2 = logo.bold.red;

  var self = this;

  self.log = function(msg,time){
    if (!cfg.verbose) return;
    if (!time) time = '';
    console.log(s, msg, time.cyan);
  };

  self.warn = function(msg){
    if (!cfg.verbose) return;
    console.warn(s2, msg);
  };
}

function readContents (location, req, res) {
  var dir = fs.readdirSync(location)
    , showIndex = req.query.houston;

  if (~dir.indexOf('index.html') && !showIndex)
    return pipeStream(path.join(location, 'index.html'), req, res);
  
  var html = readDir(location), code = 200;
  if (~html.search('Unauthorized')) code = 401;
  var stats = {
    size: html.length,
    mtime: new Date(runningTime)
  };
  if (!checkEtags(req, res, stats)) {
    res.statusCode = code;
    res.setHeader('Content-type', 'text/html');
    return res.end(html);
  }
}

//
// readDir
// Depending on the mode this function is able to read the dir
// and return the html with the files for that dir
//
function readDir (location) {
  if (!cfg.safe || !cfg.cli) {
    return '<div style="text-align:center"><h1> 401 - Unauthorized </h1><hr>' +
      '<p> node-' + process.version + ' running on ' + cfg.port + '</p></div>' ;
  }

  var dir = location
    , start = '<ul class="list"><li class="return"><a href="#" id="return">return...</a></li>'
    , listOfFiles = start + '<li><h2> Empty Dir</h2></li></ul>';

  if (typeof location !== 'object') {
    dir = fs.readdirSync(location);
  }
  if (dir.length >= 1){
    listOfFiles = start;
    listOfFiles += dir.sort().map(function(file){
      if (file[0]!=='.'){
        var relative = path.join(location, file).replace(cfg.path,'');
        var absolute = path.join(cfg.path, relative);
        if (existsSync(absolute)){
          var classEl = path.extname(absolute).substr(1);
          if (fs.statSync(absolute).isDirectory()){
              classEl = "dir";
          }
          if (relative[0] == '/') relative = relative.substr(1);
          return '<li class="'+ classEl+'"><a href="/'+relative+'">'+file+'</a></li>';
        } else {
          return '';
        }
      } else return '';
    }).join('');
    listOfFiles += '</ul>';
  } 
  return template.replace(/\{list\}/gi, listOfFiles)
                 .replace(/\{title\}/gi, location)
                 .replace(/\{version\}/gi, process.version)
                 .replace(/\{hversion\}/gi, version);
}

//
// pipeStream
// @api private
// Pipe stuff is way better than load the file to memory and then serve
// Of course that this is not a heavy-load server, but is good to keep good 
// practices
//
function pipeStream (dir, req, res) {
  function er (){
    res.writeHeader(500, {'Content-type': 'text/plain'});
    res.end('500 :: Internal server error');
  }
  fs.stat(dir, function (err, stats){
    if (err) return err;  
    if (!checkEtags(req, res, stats)){
      var stream = fs.createReadStream(dir);
      stream.on('error',er);
      stream.once('fd', function () {
        res.statusCode = 200;
        res.setHeader('Content-type', mime.lookup(dir));
       });
      stream.pipe(res);
    }
  });
}

//
// mergeObj
// merge the target object into the origin
// This overwrite the origin object
//
function mergeObj (org, target) {
  var result = org;
  Object.keys(target).map(function(prop){
    result[prop] = target[prop];
    return true;
  });
  return result;
}

//
// Partially borrowed from Connect
// lib/middleware/static
//
function checkEtags (req, res, stat) {
  function getHeader (name){
    if (req.headers[name] || req.headers[name.toLowerCase()]) return req.headers[name];
    return false;
  }
  var etag =  '"' + stat.size + '-' + Number(stat.mtime) + '"'
    , last = stat.mtime.toUTCString();
  cfg.maxAge = cfg.maxAge || 1000*3600*24;
  var ret = etag == getHeader('if-none-match') && last == getHeader('if-modified-since');
  if (!getHeader('Date')) res.setHeader('Date', new Date().toUTCString());
  if (!getHeader('Cache-Control')) res.setHeader('Cache-Control', 'public, max-age=' + (cfg.maxAge / 1000));
  if (!getHeader('Last-Modified')) res.setHeader('Last-Modified', last);
  if (!getHeader('ETag')) res.setHeader('ETag', etag);
  if (ret) {
    res.statusCode = 304;
    res.end();
    return true;
  } else {
    return false;  
  }
}
// 
// server
// The server handler
//
function server (req, res) {
  var self = this;
  req.startTime = Date.now();
  req.query  = qs.parse(url.parse(req.url).query);
  // Hack to make it work with weird paths and encoding types, especial
  // File names with spaces like "Hi world.pdf" or "niña.pdf"
  // Hat tip to @mathias
  req.url =  decodeURIComponent(url.parse(req.url).pathname);
  //
  // Emit the method to the parent if I can't handle it yep before of anything else.
  // So you can do:
  // var myapp = require('nhouston')(options);
  // myapp.on('POST', function(req,res){ /* my codez */});
  //
  if (req.method != 'GET' && self.listeners(req.method).length) {
    return self.emit(req.method, req, res);
  } else if (req.method != 'GET'){
    res.statusCode = 501;
    return res.end('Not Implemented');
  }

  var absolute = path.join(cfg.path, req.url)
    , isPublic = req.url.substr(0,8) == '/public/'
    , rabpub = ''
    , notFound = '<div style="text-align:center"><h1> 404 - Not found </h1><hr>' +
                 '<code> node-' + process.version + ' running on ' + cfg.port + '</code></div>';

  req.isPublic = isPublic;

  req.once('error', function(){
    res.writeHeader(500, {'Content-type':'text/plain'});
    res.end(notFound.replace('404 - Not found','500 - Internal Server Error'));
  });
  if (req.url == '/favicon.ico')
    return pipeStream(__dirname + '/../public/img.png', req, res);
  if (isPublic) rabpub = path.join(__dirname, '..', req.url);

  if (req.url === '/'){
    readContents(cfg.path, req, res);
  } else if (req.url !=='/favicon.ico' && !isPublic){

    if (existsSync(absolute)) {
      var stats = fs.statSync(absolute);
      if (res.statusCode == 304) return res.end();
      if (stats.isDirectory()){
        readContents(absolute, req, res);
      } else {
        pipeStream(absolute, req, res);
      }
      csl.log(req.method + req.url, (Date.now() - req.startTime) + 'ms');
    } else {
      res.writeHeader(404, {'Content-type':'text/html'});
      res.end(notFound);
    }
  } else if (isPublic && existsSync(rabpub) && !fs.statSync(rabpub).isDirectory()) {

    pipeStream(rabpub, req, res);

  } else if (isPublic && existsSync(req.url)) {

    if (path.extname(req.url) === '') {
      res.writeHeader(200, {'Content-type':'text/html'});
      return res.end(readDir(rabpub));
    }
  } else if (isPublic) {
    pipeStream(__dirname + '/../public/default.png', req, res);
  } else {
    res.writeHeader(404, {'Content-type': 'text/html'});
    res.end(notFound);
  }
}

//
// Houston
// The main function
//
function houston (options) {
  var app;

  if (!options) throw new Error('Invalid options');

  if (typeof options === 'string') cfg.path = options;
  else if (typeof options === 'object') cfg = mergeObj(cfg, options);
  else throw new TypeError('Type of ' + options + ' is not an object or a string');
  
  // Right after i read the conf launch the Logger (to be aware of verbose mode);
  csl = new Logger();

  // OH! We have a cli app, read the confs from launch args
  if (cfg.cli) load();
  // Switch protocols depending on options
  // SPDY - HTTPS - HTTP
  // And yes just require the valid one
  if (options.spdy) app = require('spdy').createServer(https.Server, options.spdy, server);
  else if (options.https) app = https.createServer(options.https, server);
  else app = http.createServer(server);
  
  if (options.spdy) {
    if (!app.listeners('request').length) {
      app.on('request', function(req, res){
        res.end('ok')
      })
    }
  }
  // Avoid loggin twice, because of domains
  var logged = 0;

  // When the server is already listing on a port log the conf
  function listener () {
    if (logged == 1) return;
    logged = 1;
    cfg.port = app.address().port;
    console.log('Server running on port := ' + cfg.port );
    // Log the current configuration
    for (var key in cfg){
      if (key != 'spdy' || key != 'https')
      csl.log(key +' : ' + cfg[key]);
    }

    // If browser is true, launch it.
    if (cfg.browser){
      var browser;
      switch (process.platform) {
        case "win32": browser = "start"; break;
        case "darwin": browser = "open"; break;
        default: browser = "xdg-open"; break;
      }
      csl.warn('Opening a new browser window...');
      require('child_process').spawn(browser, ['http://localhost:' + app.address().port ]);
    } 
  }

  // Add the hability to pick the PORT in the env var.
  if (!cfg.ignorePort) cfg.ignorePort = true;
  if (process.env.PORT && cfg.ignorePort) cfg.port = process.env.PORT;

  var handler = domain.create();
  // Yeah just once
  handler.once('error', function(){
    csl.warn('The port that you provide is in use, trying one more time');
    app.listen('', listener);
  });
  
  handler.run(function(){
    app.listen(cfg.port, listener);
  });

  // Add config (useful for cli apps)
  app.config = cfg;
  // Important config to share to parent
  app.port = cfg.port;
  app.path = cfg.path;
  // Returns a http server instance
  return app;
}

// export ALL the things
module.exports = houston;